lib/codegen.ml

```diff
@@ -0,0 +1,303 @@
+        in
+        G.fold_succ_e gen_recv g st acc
+  in
+  let callbacks = G.fold_vertex f g [] in
+  let callbacks = List.rev_map ~f:(Sig.value ~loc) callbacks in
+  let env_type = Sig.type_ Nonrecursive [Type.mk (Location.mknoloc "t")] in
+  let callbacks = Mty.signature (env_type :: callbacks) in
+  Str.modtype (Mtd.mk ~typ:callbacks (Location.mknoloc "Callbacks"))
+
+let find_all_payloads g =
+  let f (_, a, _) acc =
+    match a with
+    | SendA (_, msg) | RecvA (_, msg) -> (
+        let payloads = message_payload_ty msg in
+        match payloads with
+        | [] -> S.add acc "unit"
+        | _ -> List.fold ~f:S.add ~init:acc payloads )
+    | _ -> failwith "Impossible"
+  in
+  G.fold_edges_e f g (S.singleton (module String) "string") |> S.to_list
+
+let find_all_roles g =
+  let f (_, a, _) acc =
+    match a with
+    | SendA (r, _) | RecvA (r, _) -> S.add acc r
+    | _ -> failwith "Impossible"
+  in
+  G.fold_edges_e f g (S.empty (module String)) |> S.to_list
+
+let gen_comms_typedef payload_types =
+  let mk_recv payload_ty_str =
+    let payload_ty = mk_constr payload_ty_str in
+    let field_ty = Typ.arrow Nolabel unit payload_ty in
+    let field_name = "recv_" ^ payload_ty_str in
+    Type.field (Location.mknoloc field_name) field_ty
+  in
+  let mk_send payload_ty_str =
+    let payload_ty = mk_constr payload_ty_str in
+    let field_ty = Typ.arrow Nolabel payload_ty unit in
+    let field_name = "send_" ^ payload_ty_str in
+    Type.field (Location.mknoloc field_name) field_ty
+  in
+  let send_functions = List.map ~f:mk_send payload_types in
+  let recv_functions = List.map ~f:mk_recv payload_types in
+  let comms =
+    Type.mk
+      ~kind:(Ptype_record (send_functions @ recv_functions))
+      (Location.mknoloc "comms")
+  in
+  (* let s = [%stri type 'a t = {x: 'a}] in let s = [s] in let migrate =
+     Migrate_parsetree.Versions.migrate Migrate_parsetree.Versions.ocaml_407
+     Migrate_parsetree.Versions.ocaml_current in let s =
+     migrate.copy_structure s in Printast.structure 0 (Format.std_formatter)
+     s; *)
+  Str.type_ Nonrecursive [comms]
+
+let gen_role_ty roles =
+  let f role = Rtag (Location.mknoloc role, [], true, []) in
+  let role_ty = Typ.variant (List.map ~f roles) Closed None in
+  role_ty
+
+let gen_run_expr start g =
+  let get_transitions st =
+    let f (_, a, next) acc =
+      match a with
+      | SendA (r, msg) | RecvA (r, msg) ->
+          (r, message_label msg, message_payload_ty msg, next) :: acc
+      | _ -> failwith "Impossible"
+    in
+    G.fold_succ_e f g st []
+  in
+  let mk_run_state_name st = sprintf "run_state_%d" st in
+  let mk_run_state_ident st = Exp.ident (mk_lid (mk_run_state_name st)) in
+  let f st acc =
+    let run_state_expr =
+      match state_action_type g st with
+      | `Send ->
+          let send_fn_name = sprintf "state%dSend" st in
+          let send_callback = Exp.ident (mk_lid send_fn_name) in
+          let transitions = get_transitions st in
```

Comment:
this is used in every case, perhaps it can be moved up? Refactorings in this spirit